pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertOffHeapMetadataManager.java

```diff
@@ -0,0 +1,253 @@
+        updateRecordInfo(primaryKeyId, newRecordLocationRef);
+      }
+    }
+  }
+
+  private int getPrimaryKeyId(RecordInfo recordInfo) {
+    byte[] pk = recordInfo.getPrimaryKey().asBytes();
+    int primaryKeyId = _bytesOffHeapMutableDictionary.index(pk);
+    return primaryKeyId;
+  }
+
+  /**
+   * Updates the upsert metadata for a new consumed record in the given consuming segment.
+   */
+  public void addRecord(IndexSegment segment, RecordInfo recordInfo) {
+    ThreadSafeMutableRoaringBitmap validDocIds = Objects.requireNonNull(segment.getValidDocIds());
+    int segmentId = _segmentToSegmentIdMap.computeIfAbsent(segment, (segmentObj) -> {
+      Integer newSegmentId = _segmentId.incrementAndGet();
+      _segmentIdToSegmentMap.put(newSegmentId, segment);
+      return newSegmentId;
+    });
+
+    int primaryKeyId = getPrimaryKeyId(recordInfo);
+
+    RecordLocationRef value = getRecordInfo(primaryKeyId);
+
+    if (value != null) {
+      RecordLocationRef currentRecordLocation = value;
+      if (recordInfo.getComparisonValue().compareTo(currentRecordLocation.getComparisonValue()) >= 0) {
+        IndexSegment currentSegment = (IndexSegment) _segmentIdToSegmentMap.get(currentRecordLocation.getSegmentRef());
+        int currentDocId = currentRecordLocation.getDocId();
+        if (segment == currentSegment) {
+          validDocIds.replace(currentDocId, recordInfo.getDocId());
+        } else {
+          Objects.requireNonNull(currentSegment.getValidDocIds()).remove(currentDocId);
+          validDocIds.add(recordInfo.getDocId());
+        }
+        RecordLocationRef newRecordLocationRef =
+            new RecordLocationRef(segmentId, recordInfo.getDocId(), recordInfo.getComparisonValue());
+        updateRecordInfo(primaryKeyId, newRecordLocationRef);
+      } else {
+        updateRecordInfo(primaryKeyId, currentRecordLocation);
+      }
+    } else {
+      validDocIds.add(recordInfo.getDocId());
+      RecordLocationRef recordLocationRef =
+          new RecordLocationRef(segmentId, recordInfo.getDocId(), recordInfo.getComparisonValue());
+      updateRecordInfo(primaryKeyId, recordLocationRef);
+    }
+  }
+
+  /**
+   * Removes the upsert metadata for the given immutable segment. No need to remove the upsert metadata for the
+   * consuming segment because it should be replaced by the committed segment.
+   */
+  public void removeSegment(IndexSegment segment) {
+    throw new NotImplementedException();
+  }
+
+  //TODO: How to update data for primary key at a specific index
+  private void updateRecordInfo(int primaryKeyId, RecordLocationRef recordLocationRef) {
+    _readWriteLock.writeLock().lock();
+    _mutableForwardIndex.setInt(primaryKeyId, 0, recordLocationRef.getSegmentRef().intValue());
+    _mutableForwardIndex.setInt(primaryKeyId, 1, recordLocationRef.getDocId());
+    _mutableForwardIndex.setLong(primaryKeyId, 2, recordLocationRef.getComparisonValue());
```

Comment:
Comparison value might not always be `long`. We might need to use var-length forward index if it is not fixed-size

Reply:
What might not work? Also the comparable thing is already mentioned as a TODO. I think we can pass that as part of constructor since we can determine the data type of comparison column from schema.